### DISCUSSION TITLE
setting `OutputPath` in a `Directory.Build.targets`

### DISCUSSION THREAD

**Discussion by NekkoDroid** [dotnet/msbuild#7669] [Q&A] [answered]
When working on a plugin based project I set the default `OutputPath` to somewhere in the project root. Plugins set the `EnableDynamicLoading` property to `true` so I check that in the `Directory.Build.targets` and if it is set I override the default `OutputPath` to `$(OutputPath)plugins/`. 

**This is at least what I planned on doing...**

But neither properties are overriden/respected in the `Directory.Build.targets` while if I put it in `Directory.Build.props` the `OutputPath` is respected, but when using the Condition `'$(EnableDynamicLoading)' == 'true'` it evaluates to `false` since `EnableDynamicLoading` is only set in the plugins `.csproj` so it doesn't override the `OutputPath` for the plugins.

Now I was wondering if this is intended behavior or if this might be a bug/oversight?

**Comment by baronfel** [comment]
Can you drop a minimal sample? It _sounds_ like you may be running into the order of props and target evaluation not quite being what you expect, but with a minimal example we could talk in more detail about the specifics of your scenario.

**Reply by NekkoDroid** [reply]
```xml
<!-- Plugin.Template.csproj, located at ./Plugins/Template -->
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>

  <ItemDefinitionGroup>
    <ProjectReference Private="false" ExcludeAssets="runtime;native" />
  </ItemDefinitionGroup>

</Project>

<!-- ./Directory.Build.targets -->
<Project>

  <!-- ... -->

  <PropertyGroup Label="Project Output">
    <AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
    <AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
    <OutputPath>$(MSBuildThisFileDirectory)Build/</OutputPath>
  </PropertyGroup>

  <PropertyGroup Label="Plugins Output">
    <!-- Plugins should specify 'EnableDynamicLoading' and therefor it should be safe to check -->
    <OutputPath Condition="'$(EnableDynamicLoading)' == 'true'">$(OutputPath)plugins/</OutputPath>
  </PropertyGroup>

</Project>
```

This is what I am currently trying to do, since how I understand it, `Directory.Build.props` is imported after the SDK props file and before any "real code" specified in the .csproj. And if I understood it correctly `Directory.Build.targets` is added after any "real code" in the .csproj and before the SDK targets file (where I would have expected `OutputPath` to be read).

Here is a branch with what I was trying to do: https://github.com/NekkoDroid/Automata/tree/plugin-template

**Comment by Nirmal4G** [comment, marked as answer]
The `OutputPath` properties get overridden before hitting the Common targets and only at the end of Common targets, the `Directory.Build.targets` is imported. By purely looking at MSBuild evaluation, this should work but with all of the workarounds in the .NET SDK going on, it won't.

So, the only solution is to set this before the .NET SDK targets are imported (_implicitly_), which is just before `</Project>` tag on your project file.

There are two ways you can go about this...,

## Using `BeforeTargetFrameworkInferenceTargets` property

Separate the above "Project Output" and "Plugins Output" sections to a new targets file, let's say `JustBeforeSdk.targets` and then put the following in your `Directory.Build.props`. This gets called in before the `TargetFrameworkInference`, meaning that you can also specify a custom target framework and resolve it yourself in there. That was the original intent for the hook. Since, it gets imported before setting/overriding the output paths, we can use that!

```xml
<!-- The 'PathToYourBuildFiles' property should end with a trailing slash and should be an absolute path! -->
<BeforeTargetFrameworkInferenceTargets>$(PathToYourBuildFiles)JustBeforeSdk.targets</BeforeTargetFrameworkInferenceTargets>
```

**Do note that this only works if your .NET SDK is v5 and above.**

## Using an import at the end of the project file

Import `JustBeforeSdk.targets` at the end of every project file you want to modify. This is tedious one if you have a large number of projects. But this works with all versions of .NET SDKs. since, it doesn't depend on a hook inside the SDK that was only added from v5+.

```xml
<!-- Plugin.Template.csproj, located at ./Plugins/Template -->
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <EnableDynamicLoading>true</EnableDynamicLoading>
  </PropertyGroup>

  <ItemDefinitionGroup>
    <ProjectReference Private="false" ExcludeAssets="runtime;native" />
  </ItemDefinitionGroup>

  <!-- The 'PathToYourBuildFiles' should end with trailing slash, an absolute path and should be declared in your `Directory.Build.props` -->
  <Import Project="$(PathToYourBuildFiles)JustBeforeSdk.targets" />

</Project>
```

**Reply by NekkoDroid** [reply]
Thank you for your response and explanation.

Your first suggestion is what I was looking for since I want all the custom build output to be configured in a single location without needing to add an import on all projects (The plugins shouldn't need to care about where they are built to, just how to build themselves, so the second would kinda go against that).
